src/transformers/models/encoder_decoder/modeling_encoder_decoder.py

```diff
@@ -175,6 +175,12 @@ def __init__(
 
         self.encoder = encoder
         self.decoder = decoder
+        
+        # we can not stop the users from calling the encoder or decoder individually,
+        # we need to link the individual configs with the encoderdecoderconfig to ensure consistency
+        self.config.encoder = self.encoder.config
+        self.config.decoder = self.decoder.config
```

Comment:
(nit) could we actually do this other way around, think that's more cleaner. models config should refer to shared config.

Reply:
Just to break it down here, this is needed whenever no `config` is passed since in this case a new config is created from the encoder and decoder & this leads to problems right? Whenever a config is passed and **no** encoder/decoder is passed this should be fine since encoder & decoder are created from the configs. Then when both `config` and `encoder`/`decoder` are passed we should probably overwrite the `encoder`/`decoder` config. 

=> So I agree with @patil-suraj here that the better way would be to make sure `self.encoder.config = self.config.encoder`  => Also it would be great if we could add a test in the beginning the compares both configs & if they are not identical we throw a warning that `encoder.config` is overwritten with `config.encoder`.

Reply:
So I think we should do the following:

```python 
if self.encoder.config.to_dict() != self.config.encoder.to_dict():
   logging.warning(f"Config of the encoder: {self.encoder.__class__} is overwritten by shared encoder config: {self.config.encoder}")

# same for decoder

self.encoder.config = self.config.encoder
self.decoder.config = self.conig.decoder
```

Reply:
> Just to break it down here, this is needed whenever no `config` is passed since in this case a new config is created from the encoder and decoder & this leads to problems right? Whenever a config is passed and **no** encoder/decoder is passed this should be fine since encoder & decoder are created from the configs. Then when both `config` and `encoder`/`decoder` are passed we should probably overwrite the `encoder`/`decoder` config.

Exactly, the following will allow diverging configs:
```
m = EncoderDecoderModel.from_encoder_decoder_pretrained('...', '...')
m = EncoderDecoderModel(encoder=a, decoder=b)
m = EncoderDecoderModel(config=c, encoder=a, decoder=b)
```
While these will have tied configs:
```
m = EncoderDecoderModel(config=c)
m = EncoderDecoderModel.from_pretrained("....")
```